indra/llwindow/llgamecontrol.cpp

```diff
@@ -441,57 +620,144 @@ std::string LLGameControl::Options::AxisOptions::saveToString() const
         options.push_back(llformat("offset:%d", mOffset));
     }
 
-    std::string result = LLStringOps::join(options);
+    std::string result = LLStringUtil::join(options);
 
     return result.empty() ? result : "{" + result + "}";
 }
 
-std::string LLGameControl::Options::saveToString() const
+// Parse string "{key:value,key:{key:value,key:value}}" and fill the map
+static bool parse(std::map<std::string, std::string>& result, std::string source)
 {
-    std::list<std::string> options;
+    result.clear();
 
-    auto pair2str = [](const std::pair<U8, U8>& pair) -> std::string
-    {
-        return pair.first == pair.second ? LLStringUtil::null : llformat("%u:%u", pair.first, pair.second);
+    LLStringUtil::trim(source);
+    if (source.empty())
+        return true;
+
+    if (source.front() != '{' || source.back() != '}')
+        return false;
+
+    source = source.substr(1, source.size() - 2);
+
+    LLStringUtil::trim(source);
+    if (source.empty())
+        return true;
+
+    // Split the string "key:value" and add the pair to the map
+    auto split = [&](const std::string& pair) -> bool
+    {
+        size_t pos = pair.find(':');
+        if (!pos || pos == std::string::npos)
+            return false;
+        std::string key = pair.substr(0, pos);
+        std::string value = pair.substr(pos + 1);
+        LLStringUtil::trim(key);
+        LLStringUtil::trim(value);
+        if (key.empty() || value.empty())
+            return false;
+        result[key] = value;
+        return true;
     };
 
-    std::string axis_map = LLStringOps::join<std::map<U8, U8>, std::pair<U8, U8>>(mAxisMap, pair2str);
-    if (!axis_map.empty())
+    U32 depth = 0;
+    size_t offset = 0;
+    while (true)
     {
-        options.push_back("axis_map:{" + axis_map + "}");
+        size_t pos = source.find_first_of(depth ? "{}" : ",{}", offset);
+        if (pos == std::string::npos)
+        {
+            return !depth && split(source);
+        }
+        if (source[pos] == ',')
+        {
+            if (!split(source.substr(0, pos)))
+                return false;
+            source = source.substr(pos + 1);
+            offset = 0;
+        }
+        else if (source[pos] == '{')
+        {
+            depth++;
+            offset = pos + 1;
+        }
+        else if (depth) // Assume '}' here
+        {
+            depth--;
+            offset = pos + 1;
+        }
+        else
+        {
+            return false; // Extra '}' found
+        }
     }
```

Comment:
Since you're performing the operation `depth--`, that variable should be `S32` instead of `U32` , and the loop should `return false` immediately when `depth` goes negative.